ThermoViscoProblem.py

```diff
@@ -344,7 +358,61 @@ def _setup_solver(self) -> None:
         opts[f"{option_prefix}pc_type"] = "gamg"
         opts[f"{option_prefix}pc_factor_mat_solver_type"] = "mumps"
         self.ksp.setFromOptions()
+        
+    # linear elasticity equation #
+
     
+    def _set_dirichlet_bc(self) -> None:
+        
+        u_bc = fem.Constant(self.mesh, PETSc.ScalarType((0.0,0.0)))
+        facet_dim = self.mesh.topology.dim-1
+        facets_l = locate_entities_boundary(self.mesh, dim=facet_dim, marker=lambda x: np.isclose(x[0], 0.0))
+        facets_r = locate_entities_boundary(self.mesh, dim=facet_dim, marker=lambda x: np.isclose(x[0], 50.0))
+        facets_t = locate_entities_boundary(self.mesh, dim=facet_dim, marker=lambda x: np.isclose(x[1], 0.0))
+        facets_b = locate_entities_boundary(self.mesh, dim=facet_dim, marker=lambda x: np.isclose(x[1], 25.0))
+        dofs_l = fem.locate_dofs_topological(V=self.functionSpaces["U"], entity_dim=facet_dim, entities=facets_l)
+        dofs_r = fem.locate_dofs_topological(V=self.functionSpaces["U"], entity_dim=facet_dim, entities=facets_r)
+        dofs_t = fem.locate_dofs_topological(V=self.functionSpaces["U"], entity_dim=facet_dim, entities=facets_t)
+        dofs_b = fem.locate_dofs_topological(V=self.functionSpaces["U"], entity_dim=facet_dim, entities=facets_b)
+        self.bc = [#fem.dirichletbc(u_bc, dofs_l, self.functionSpaces["U"]),
+                   #fem.dirichletbc(u_bc, dofs_r, self.functionSpaces["U"]),
+                   #fem.dirichletbc(u_bc, dofs_t, self.functionSpaces["U"]),
+                   fem.dirichletbc(u_bc, dofs_b, self.functionSpaces["U"])
+                   ]
+    
+    def _setup_weak_form_u(self) -> None:
+        
+        ds = Measure("exterior_facet",domain=self.mesh)
+        dx = Measure("dx",domain=self.mesh)
+        
+        self.ss = Constant(self.mesh,default_scalar_type((0,-3.0)))        # Body force
+        self.traction = Constant(self.mesh,default_scalar_type((0.0,0.0)))   # traction force
+        
+        # Lame's elasticity parameters
+        
+        mu = self.material_model.mu 
+        lambda_ = self.material_model.lambda_  
+        I = self.material_model.I
+        
+        
+        def elastic_epsilon(ua):
+            return sym(grad(ua)) 
+
+        def elastic_sigma(ua):
+            return lambda_ * (nabla_div(ua))* I + 2 * mu * elastic_epsilon(ua)
+        
+        self.a = inner(elastic_sigma(self.u_trial), elastic_epsilon(self.v_test)) * dx 
+        self.L = dot(self.ss, self.v_test) * dx + dot(self.traction,self.v_test) * ds
+
+        self.expressions = {}
```

Comment:
I tried to write and recall this expression function from ViscoelasticModel.py, but elastic_epsilon() could not be accessed as it is function inside a function.

Reply:
Yeah, I get the problem. Maybe it makes sense to refactor `elastic_sigma` and `elastic_epsilon` into `ViscoelasticModel`. We can then access all functions and the expression via `self.material_model.elastic_sigma`.

Another thing I noticed: you set the key for the expression in line 408 to `"U"`, but isn't that the mechanical (elastic) strain? If so, it would make more sense to create a key `"elastic_strain"` since we called the thermal strain `"thermal_strain"` within the expressions dictionary:
https://github.com/PI-UniA/SurroGlas/blob/7356bd8183be36dd5c7b1187c1fb8519a1a78734/ViscoelasticModel.py#L127-L133